## Exploring Ebay Car Sales Data.
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
you can find the link of the dataset : https://data.world/data-society/used-cars-data.
The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

> The aim of this project is to clean the data and analyze the included used car listings

In [1]:
import pandas as pd 
import numpy as np 

In [2]:
autos = pd.read_csv("autos.csv",encoding = "Latin-1")

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
print(type(autos))
print(autos.shape)

<class 'pandas.core.frame.DataFrame'>
(50000, 20)


In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

> - The dataset contains 20 columns, most of which are strings.
> - Some columns have null values, but none have more than ~20% null values.
>- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

### 2- Cleaning Column Names

In [7]:
autos.rename(columns=
             {"yearOfRegistration":"registration_year","monthOfRegistration":"registration_month",
              "notRepairedDamage":"unrepaired_damage","dateCreated":"ad_created","dateCrawled":"date_Crawled",
              "offerType":"offer_Type","vehicle_Type":"vehicle_Type","fuelType":"fuel_Type","nrOfPictures":"nr_Of_Pictures",
             "postalCode":"postal_Code","lastSeen":"last_Seen"},inplace=True)

> Any columns that have mostly one value that are candidates to be dropped<br>
> Any columns that need more investigation.<br>
> Any examples of numeric data stored as text that needs to be cleaned and converted.

### 3- Inital Exploration and Cleaning.

In [8]:
autos =  autos.drop(["nr_Of_Pictures","seller","offer_Type"],axis= 1)
print(autos.shape)
print(autos.columns)

(50000, 17)
Index(['date_Crawled', 'name', 'price', 'abtest', 'vehicleType',
       'registration_year', 'gearbox', 'powerPS', 'model', 'odometer',
       'registration_month', 'fuel_Type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_Code', 'last_Seen'],
      dtype='object')


In [9]:
autos["price"]= autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [10]:
autos.rename(columns={"odometer":"odometer_km"},inplace=True)

In [11]:
print(autos["odometer_km"].unique().shape) 
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


> - 13 unique value in "odmeter_km".
> - the maximume odometer_km 150000 with frequncey 32424 counts.\
> - the minimume odometer_km 5000 with frequancey 967 counts.

### 4- Exploring the Odometer and Price Columns

In [12]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().head(10))
print(autos["price"].max())
print(autos["price"].min())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64
99999999
0


In [13]:
autos["price"].value_counts().sort_index(ascending = True)

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

There are a number of listings with prices below \\$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over \\$1 million.  

Given that eBay is an auction site, there could legitimately be items where the opening bid is \\$1. We will keep the \\$1 items, but remove anything above \\$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.<br>
## Removing outliers

In [14]:
autos = autos[autos["price"].between(1,350000)]
autos.head()

date_Crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicleType  registration_year    gearbox  powerPS   model  \
0   5000  control         bus               2004    manuell      158  andere   
1   8500  control   limousine               1997  automatik      286     7er   
2   8990     test   limousine               2009    manuell      102    golf   
3   4350  control  kleinwagen               2007  automatik       71  fortwo   
4   1350     test       kombi               2003    manuell        0   focus   

   odometer_km  registration_month fuel_Type       brand unrepaired_damage  \
0       150000                   3       lpg     peugeot              nein   
1       150000                   6    benzin         bmw              nein   
2        70000                   7    benzin  volkswagen              nein   
3        70000                   6    benzin       smart              nein   
4       150000                   7    benzin        ford              nein   

            ad_created  postal_Code            last_Seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [15]:
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

some prices are a lot and over expinssive so we removed them.

   Let's now move on to the date columns and understand the date range the data covers.

### 5-Exploring The Data Columns

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively.

In [16]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_Crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price               48565 non-null  int32 
 3   abtest              48565 non-null  object
 4   vehicleType         43979 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gearbox             46222 non-null  object
 7   powerPS             48565 non-null  int64 
 8   model               46107 non-null  object
 9   odometer_km         48565 non-null  int32 
 10  registration_month  48565 non-null  int64 
 11  fuel_Type           44535 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   39464 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_Code         48565 non-null  int64 
 16  last_Seen           48

In [17]:
autos[["date_Crawled","ad_created","last_Seen"]].describe()

date_Crawled           ad_created            last_Seen
count                 48565                48565                48565
unique                46882                   76                38474
top     2016-03-23 19:38:20  2016-04-03 00:00:00  2016-04-07 06:17:27
freq                      3                 1887                    8

In [21]:
#calculate the distribution of values as percentage including the including the null values
print(autos["date_Crawled"].value_counts(normalize=True, dropna=False))
print("\n")
print(autos["ad_created"].value_counts (normalize=True, dropna=False))
print("\n")
print(autos["last_Seen"].value_counts(normalize=True, dropna=False))

2016-03-23 19:38:20    0.000062
2016-04-02 11:37:04    0.000062
2016-03-12 16:06:22    0.000062
2016-04-02 15:49:30    0.000062
2016-03-23 18:39:34    0.000062
                         ...   
2016-03-07 18:55:38    0.000021
2016-03-09 11:53:12    0.000021
2016-03-23 20:37:20    0.000021
2016-04-03 17:49:23    0.000021
2016-03-16 18:52:33    0.000021
Name: date_Crawled, Length: 46882, dtype: float64


2016-04-03 00:00:00    0.038855
2016-03-20 00:00:00    0.037949
2016-03-21 00:00:00    0.037579
2016-04-04 00:00:00    0.036858
2016-03-12 00:00:00    0.036755
                         ...   
2016-02-16 00:00:00    0.000021
2016-01-07 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
2016-01-03 00:00:00    0.000021
2016-01-14 00:00:00    0.000021
Name: ad_created, Length: 76, dtype: float64


2016-04-07 06:17:27    0.000165
2016-04-07 03:16:17    0.000144
2016-04-06 21:17:51    0.000144
2016-04-06 15:45:50    0.000124
2016-04-06 02:16:12    0.000124
                         ...   
2016-

In [22]:
print(autos["date_Crawled"].sort_index())
print(autos["ad_created"].sort_index())
print(autos["last_Seen"].sort_index())

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
                ...         
49995    2016-03-27 14:38:19
49996    2016-03-28 10:50:25
49997    2016-04-02 14:44:48
49998    2016-03-08 19:25:42
49999    2016-03-14 00:42:12
Name: date_Crawled, Length: 48565, dtype: object
0        2016-03-26 00:00:00
1        2016-04-04 00:00:00
2        2016-03-26 00:00:00
3        2016-03-12 00:00:00
4        2016-04-01 00:00:00
                ...         
49995    2016-03-27 00:00:00
49996    2016-03-28 00:00:00
49997    2016-04-02 00:00:00
49998    2016-03-08 00:00:00
49999    2016-03-13 00:00:00
Name: ad_created, Length: 48565, dtype: object
0        2016-04-06 06:45:54
1        2016-04-06 14:45:08
2        2016-04-06 20:15:37
3        2016-03-15 03:16:28
4        2016-04-01 14:38:50
                ...         
49995    2016-04-01 13:47:40
49996    2016-04-02 14:18:02
49997    2016-04-04 11:47:27
49998

In [23]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

> One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:<br>
 The minimum value is 1000, before cars were invented<br>
 The maximum value is 9999, many years into the future
 
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

### 6- Dealing With Incorrect Regstration Year Data.
Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely

In [24]:
print(autos["registration_year"].unique())
print("\n")
print(autos["registration_year"].value_counts())

[2004 1997 2009 2007 2003 2006 1995 1998 2000 2017 2010 1999 1982 1990
 2015 2014 1996 1992 2002 2012 2011 2005 2008 1985 2016 1994 1986 2001
 2018 2013 1972 1993 1988 1989 1973 1967 1976 4500 1987 1991 1983 1960
 1969 1950 1978 1980 1984 1963 1977 1961 1968 1934 1965 1971 1966 1979
 1981 1970 1974 1910 1975 5000 4100 2019 1956 9999 6200 1964 1959 1958
 1800 1948 1931 1943 1941 1962 1927 1937 1929 1000 1957 1952 1111 1955
 1939 8888 1954 1938 2800 5911 1953 1951 4800 1001 9000]


2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
        ... 
9000       1
1948       1
1000       1
1931       1
1952       1
Name: registration_year, Length: 95, dtype: int64


In [25]:
# for the minimum will be 1900 and the maximum will 2016.
#autos["registration_year"].head()
autos = autos[(autos["registration_year"] <= 2016)&(autos["registration_year"] >= 1900)]

In [26]:
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

### 7-Exploring Price by Brand
we want to select those that have over a certain percentage of the total values e.g. > 5%

In [27]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [28]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


### 8- Storing Aggregate Data in a DataFrame

In [29]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}


Of the top 5 brands, there is a distinct price gap:

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option

### Exploring Mileage 

In [30]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [32]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [33]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [34]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

> The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.